grpc/src/jmh/java/com/linecorp/armeria/grpc/shared/SimpleBenchmarkBase.java

```diff
@@ -0,0 +1,195 @@
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.protobuf.Empty;
+
+import com.linecorp.armeria.benchmarks.GithubApi.SearchResponse;
+import com.linecorp.armeria.benchmarks.GithubServiceGrpc;
+import com.linecorp.armeria.benchmarks.GithubServiceGrpc.GithubServiceBlockingStub;
+import com.linecorp.armeria.benchmarks.GithubServiceGrpc.GithubServiceFutureStub;
+
+import io.grpc.ManagedChannel;
+import io.grpc.okhttp.OkHttpChannelBuilder;
+
+@State(Scope.Benchmark)
+public abstract class SimpleBenchmarkBase {
+
+    @AuxCounters
+    @State(Scope.Thread)
+    public static class Counters {
+        private AtomicLong numRequests = new AtomicLong();
+        private AtomicLong numFailures = new AtomicLong();
+        private AtomicLong currentRequests = new AtomicLong();
+
+        private volatile boolean waiting;
+
+        public long numRequests() {
+            return numRequests.get();
+        }
+
+        public long numFailures() {
+            return numFailures.get();
+        }
+
+        public long currentRequests() {
+            return currentRequests.get();
+        }
+
+        @Setup(Level.Iteration)
+        public void reset() {
+            waiting = false;
+            numRequests.set(0);
+            numFailures.set(0);
+            currentRequests.set(0);
+        }
+
+        @TearDown(Level.Iteration)
+        public void waitForCurrentRequests() {
+            waiting = true;
+            await().forever().until(() -> currentRequests.get() == 0);
+        }
+    }
+
+    /**
+     * The port the benchmark's server is listening on.
+     */
+    protected abstract int port();
+
+    /**
+     * The normal {@link GithubServiceBlockingStub} for the benchmark. The okhttp version will be set up in this
+     * class as it's the same for both upstream and downstream.
+     */
+    protected abstract GithubServiceBlockingStub normalClient();
+
+    /**
+     * The normal {@link GithubServiceFutureStub} for the benchmark. The okhttp version will be set up in this
+     * class as it's the same for both upstream and downstream.
+     */
+    protected abstract GithubServiceFutureStub normalFutureClient();
+
+    /**
+     * Benchmark initialization logic.
+     */
+    protected abstract void setUp() throws Exception;
+
+    /**
+     * Benchmark teardown logic.
+     */
+    protected abstract void tearDown() throws Exception;
+
+    private ManagedChannel okhttpChannel;
+    private GithubServiceBlockingStub githubApiOkhttpClient;
+    private GithubServiceFutureStub githubApiOkhttpFutureClient;
+
+    @Param
+    public ClientType clientType = ClientType.NORMAL;
```

Comment:
- Can be private
- No need to set the initial value

Reply:
Indeed it does - it seems to complain about private on everything except for fields, that's surprising...